### PR TITLE
Better script for prepare release notes

### PR DESCRIPTION
**Pull request by Czaki**
# Description

Based on discussions from meetings, I decided to improve the Release notes generation script.

What is improved:
1) For pointed tag and branch, the newest ancestor commit is found and all commits since ancestor are collected. Then commits from branch to release are excluded based on PR number form the previous release branch (I expect that the pattern from 0.4.16 may repeat).  
2) collect all labels related with PR and assign them to category base on labels (need to be improved) 
3) Introduced cache for GitHub API requests. Speedup for a normal run, a significant help for development. 

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

@andy-sweet as the current release manager, you maybe interested. release notes are generated with the script.

### CODE REVIEW

File: docs/release/generate_release_notes.py
Comment (Czaki) [inline docs/release/generate_release_notes.py:250]: It could be speedup using threads, but I'm not sure if worth it.

File: docs/release/generate_release_notes.py
Comment (Czaki) [inline docs/release/generate_release_notes.py]: when updating labels on GitHub, there is a requirement to remove the cache file. Maybe it should be handled somehow better?

File: docs/release/generate_release_notes.py
Comment (Czaki) [inline docs/release/generate_release_notes.py:89]: I do not like this, but do not have a better idea.

File: docs/release/generate_release_notes.py
Comment (Czaki) [inline docs/release/generate_release_notes.py:145]: For proper working, the local main need to be the same as the remote one. But maybe one knows how to specify remote labels?

File: docs/release/generate_release_notes.py
Comment (Czaki) [inline docs/release/generate_release_notes.py:126]: this requires that the target branch is up to date locally.

**Review by Carreau — APPROVED** [review]

**Comment by Carreau** [comment]
Not for this PR, but the thing that would be great would be to also automatically put the right milestone on the PRs that don't have milestone.

File: docs/release/generate_release_notes.py
Comment (andy-sweet) [inline docs/release/generate_release_notes.py:104]: Is this implementation comment still relevant now?

File: docs/release/generate_release_notes.py
Comment (andy-sweet) [inline docs/release/generate_release_notes.py:130]: Is this implementation comment still relevant now?

```suggestion
```

**Review by andy-sweet — APPROVED** [review]
I'm not very familiar with the GitHub API, but this looks good to me. I was able to pull this down, run the script, and generated the same output as you did here.

**Comment by andy-sweet** [comment]
Will merge after 24 hours unless there is an objection.

File: docs/release/generate_release_notes.py
Comment (Czaki) [reply docs/release/generate_release_notes.py:130]: I still get commit from remote, not tagged one, but the common ancestor of tag and branch.

Comment (andy-sweet) [reply docs/release/generate_release_notes.py:130]: Yeah, the reason seems more clear after the new code and variable names you added. So this comment doesn't seem to add much now. But I'd let you decide if you want to keep it.

**Comment by Czaki** [comment]
What did you think about highlight PR also based on labels?

**Comment by andy-sweet** [comment]
> What did you think about highlight PR also based on labels?

How would you determine what a highlight is? Are there specific labels that would classify them? Maybe `Epic`?

My intuition is to avoid automatically classify highlights. I feel like they are more like a manually curated small set of features/improvements that we want to bring attention to and maybe give some context/information on. There might be some bias in how they are selected, but I think anyone can propose adding a highlight.

**Comment by Czaki** [comment]
> How would you determine what a highlight is? Are there specific labels that would classify them? Maybe `Epic`?

Yes. I think about such a Label that could be added to PR during merge time to get an initial highlight set for preparing releases notes.

**Review by andy-sweet — APPROVED** [review]

**Review by jni — APPROVED** [review]
This is great @Czaki! ❤️ 

We can iterate on this, but I think it would be great to keep the labels as close to the section name as possible:

```
label_to_section = {
    "bug": "Bug Fixes",
    "bugfix": "Bug Fixes",
    "feature": "New Features",
    "api": "API Changes",
    "enhancement": "Improvements",
    "deprecation": "Deprecations",
    "dependencies": "Build Tools",
    "documentation": "Documentation",
}
```

- I think we need a section/label for UI changes, either "ui" or "gui"
- Should we rename the enhancement label to improvement?
- I don't think "dependencies" -> "build tools" makes that much sense?
- Perhaps we should add a section/label for "performance", given that it's one of our values? I think it would be great to see progress on performance happening (or not) in each release.
- Should we change all of these labels to the same colour? I think it would make sense so that core devs grow to "sense" that a PR is wrong just because that splash of colour is missing from the right sidebar. 😂 
- I agree with you that we should have a "highlight" label

I haven't checked this, but can a PR end up in more than one section? I think they should be able to do that.

Thank you!

**Comment by Czaki** [comment]
> We can iterate on this, but I think it would be great to keep the labels as close to the section name as possible:

Agree. I try to fit current sections and labels to get a useful script for the current release. I think that the next step should be to depend only on PR labels, with the selection of labels that will be used during generation (should be different from these labels that are added by the template of GH actions.). 


> I don't think "dependencies" -> "build tools" makes that much sense?

just checked such labeled PR and it looks accurate, but it may be coincidence.


> Perhaps we should add a section/label for "performance", given that it's one of our values? I think it would be great to see progress on performance happening (or not) in each release.

Yes. 

> I haven't checked this, but can a PR end up in more than one section? I think they should be able to do that.

Currently not, I'm worried about labels in issues, but could be easy changed.

**Comment by Czaki** [comment]
I have updated the behavior. The script first tries to determine sections (possible multiple assignments) based on the PR labels. 
If no section is added, then it tries to determine related issues and then based on its labels adds to some section (single assignment). 

Also, add the Highlight label and assign it to two arbitrary selected PR.

**Comment by andy-sweet** [comment]
Looks like all the comments have been addressed, except for ones related to the current GitHub labels (e.g. color, names) and the possible addition of a GUI section in the release notes. I think those can be done independently of this first PR, so will likely merge this soon.

I also pulled the latest version of this branch down and re-ran the script, but got slightly different output that the release notes file here - [see the diff for details](https://gist.github.com/andy-sweet/68b2b3e90a8765c3af3ae823f197774c). Is there something I need to run/update before re-running the script?

**Comment by Czaki** [comment]
> I also pulled the latest version of this branch down and re-ran the script, but got slightly different output that the release notes file here - [see the diff for details](https://gist.github.com/andy-sweet/68b2b3e90a8765c3af3ae823f197774c). Is there something I need to run/update before re-running the script?

you need to delete the cache for github requests, a file named: `github_cache.sqlite`

Maybe I need to add some additional logic here. I'm now trying to understand how to set up cache expiration.

**Comment by Czaki** [comment]
@andy-sweet I have added global cache with expiration 1h and 2 minute cache fro new commit list. But I'm not sure is this are proper values.

**Comment by andy-sweet** [comment]
> I have added global cache with expiration 1h and 2 minute cache fro new commit list. But I'm not sure is this are proper values.

Agreed that these are somewhat arbitrary, but given that this tool will only be used by release managers for now, I think that's fine. Will probably merge this within the next 24 hours.

**Pull request merged**
